tests/unit_tests/sparse_blas/source/sparse_spmm_usm.cpp

```diff
@@ -0,0 +1,296 @@
+    }
+
+    auto ia_usm_uptr = malloc_device_uptr<intType>(main_queue, ia_host.size());
+    auto ja_usm_uptr = malloc_device_uptr<intType>(main_queue, ja_host.size());
+    auto a_usm_uptr = malloc_device_uptr<fpType>(main_queue, a_host.size());
+    auto b_usm_uptr = malloc_device_uptr<fpType>(main_queue, b_host.size());
+    auto c_usm_uptr = malloc_device_uptr<fpType>(main_queue, c_host.size());
+    auto alpha_usm_uptr = malloc_device_uptr<fpType>(main_queue, 1);
+    auto beta_usm_uptr = malloc_device_uptr<fpType>(main_queue, 1);
+
+    intType *ia_usm = ia_usm_uptr.get();
+    intType *ja_usm = ja_usm_uptr.get();
+    fpType *a_usm = a_usm_uptr.get();
+    fpType *b_usm = b_usm_uptr.get();
+    fpType *c_usm = c_usm_uptr.get();
+
+    std::vector<sycl::event> mat_dependencies;
+    std::vector<sycl::event> spmm_dependencies;
+    // Copy host to device
+    mat_dependencies.push_back(
+        main_queue.memcpy(ia_usm, ia_host.data(), ia_host.size() * sizeof(intType)));
+    mat_dependencies.push_back(
+        main_queue.memcpy(ja_usm, ja_host.data(), ja_host.size() * sizeof(intType)));
+    mat_dependencies.push_back(
+        main_queue.memcpy(a_usm, a_host.data(), a_host.size() * sizeof(fpType)));
+    spmm_dependencies.push_back(
+        main_queue.memcpy(b_usm, b_host.data(), b_host.size() * sizeof(fpType)));
+    spmm_dependencies.push_back(
+        main_queue.memcpy(c_usm, c_host.data(), c_host.size() * sizeof(fpType)));
+
+    fpType *alpha_host_or_usm_ptr = &alpha;
+    fpType *beta_host_or_usm_ptr = &beta;
+    if (test_scalar_on_device) {
+        spmm_dependencies.push_back(
+            main_queue.memcpy(alpha_usm_uptr.get(), &alpha, sizeof(fpType)));
+        spmm_dependencies.push_back(main_queue.memcpy(beta_usm_uptr.get(), &beta, sizeof(fpType)));
+        alpha_host_or_usm_ptr = alpha_usm_uptr.get();
+        beta_host_or_usm_ptr = beta_usm_uptr.get();
+    }
+
+    sycl::event ev_copy, ev_spmm;
+    oneapi::mkl::sparse::matrix_handle_t A_handle = nullptr;
+    oneapi::mkl::sparse::dense_matrix_handle_t B_handle = nullptr;
+    oneapi::mkl::sparse::dense_matrix_handle_t C_handle = nullptr;
+    oneapi::mkl::sparse::spmm_descr_t descr = nullptr;
+    std::unique_ptr<std::uint8_t, UsmDeleter> workspace_usm(nullptr, UsmDeleter(main_queue));
+    try {
+        init_sparse_matrix(main_queue, format, &A_handle, nrows_A, ncols_A, nnz, index, ia_usm,
+                           ja_usm, a_usm);
+        for (auto property : matrix_properties) {
+            CALL_RT_OR_CT(oneapi::mkl::sparse::set_matrix_property, main_queue, A_handle, property);
+        }
+        CALL_RT_OR_CT(oneapi::mkl::sparse::init_dense_matrix, main_queue, &B_handle, opb_nrows,
+                      opb_ncols, ldb, dense_matrix_layout, b_usm);
+        CALL_RT_OR_CT(oneapi::mkl::sparse::init_dense_matrix, main_queue, &C_handle,
+                      static_cast<std::int64_t>(opa_nrows), ncols_C, ldc, dense_matrix_layout,
```

Comment:
I have tried to reduce the number of `static_cast` already but some of them are inevitable as I prefer to keep the compiler warnings related to integer overflow. I don't think I can remove the `static_cast`, I could move them to their own variable but this would only add more verbosity as the variables would be used only once. If your goal is to reduce verbosity then I don't have a good solution. Do you have something else in mind?

Reply:
Up on line 47/48, you have:
```cpp
    auto [opa_nrows, opa_ncols] = swap_if_transposed<std::size_t>(transpose_A, nrows_A, ncols_A);
    auto [opb_nrows, opb_ncols] = swap_if_transposed<std::int64_t>(transpose_B, opa_ncols, ncols_C);
```
Is it not okay to use `std::int64_t` there in the first one given you're already doing that for the latter?
When you are _up_-casting (`std::int64_t` to `size_t`), you probably don't need a `static_cast`, but you do want to indicate readers of explicit _down_-casting (`size_t` to `std::int64_t`) to warn them of possible overflows (even though we won't have them in these small unit tests of course.)

Reply:
There are 3 occurrences where `opa_nrows` and `opa_ncols` are used as `std::size_t` and only one where `opa_nrows` is used as `std::int64_t` so it was simpler to keep them as `std::size_t`.
I think your point makes sense if we assume that we don't use negative values, otherwise both casts are error-prone. That's why I went for the solution that has the fewer casts. I don't mind changing it, does it look worth it to you?

Reply:
Nah, it's not a big deal in the tests. If it were the main source code, however, this would have been worth looking into.